### PR TITLE
Implement routing for ListOffsets

### PR DESCRIPTION
**Pull request by rukai**
From inspecting the messages passing through shotover I could see that ListOffsets is hitting NOT_LEADER_OR_FOLLOWER errors. (error code 6)

For example:
```
shotover   02:07:50.445929Z  INFO connection{id=41 source="kafka"}: shotover::transforms::debug::printer: Response: Kafka version:7 correlation_id:54 ListOffsets(ListOffsetsResponse { throttle_time_ms: 0, topics: [ListOffsetsTopicResponse { name: TopicName("partitions3_case4"), partitions: [ListOffsetsPartitionResponse { partition_index: 0, error_code: 6, old_style_offsets: [], timestamp: -1, offset: -1, leader_epoch: -1, unknown_tagged_fields: {} }], unknown_tagged_fields: {} }], unknown_tagged_fields: {} })
```

After some retries it will eventually hit the correct node and succeed:
```
shotover   02:07:51.047014Z  INFO connection{id=41 source="kafka"}: shotover::transforms::debug::printer: Response: Kafka version:7 correlation_id:60 ListOffsets(ListOffsetsResponse { throttle_time_ms: 0, topics: [ListOffsetsTopicResponse { name: TopicName("partitions3_case4"), partitions: [ListOffsetsPartitionResponse { partition_index: 0, error_code: 0, old_style_offsets: [], timestamp: -1, offset: 0, leader_epoch: 0, unknown_tagged_fields: {} }], unknown_tagged_fields: {} }], unknown_tagged_fields: {} })
```
It seems that the java driver is robust enough to retry only the parts of the request that failed, so eventually it will succesfully complete and move on. However, it is still quite slow and wasteful to error like this.
To fix these errors, we need to split and combine the request like we do for fetch and produce messages.

So this PR implements that fetch/combine logic for ListOffsets.

I added an integration test to call the listOffsets method from the admin API.
This test does not fail due to java's robust retry mechanism but I've manually verified that the errors are gone with the new routing logic and the new test adds more coverage of the driver.

Another nice outcome of this fix is that `cluster_1_rack_single_shotover::case_2_java` now completes in 100s down from 110s

### CODE REVIEW

File: shotover/src/transforms/kafka/sink_cluster/mod.rs [hunk 1]
```diff
@@ -1101,6 +1105,140 @@ impl KafkaSinkCluster {
...
+    fn route_list_offsets(&mut self, mut request: Message) -> Result<()> {
+        if let Some(Frame::Kafka(KafkaFrame::Request {
+            body: RequestBody::ListOffsets(list_offsets),
+            ..
+        })) = request.frame()
+        {
+            let routing = self.split_list_offsets_request_by_destination(list_offsets);
+
+            if routing.is_empty() {
+                // Fetch contains no topics, so we can just pick a random destination.
```
Comment (rukai) [inline shotover/src/transforms/kafka/sink_cluster/mod.rs:1169]: ListOffsets

File: shotover/src/transforms/kafka/sink_cluster/mod.rs [hunk 2]
```diff
@@ -1576,6 +1718,46 @@ impl KafkaSinkCluster {
...
+        drain: impl Iterator<Item = Message>,
+    ) -> Result<()> {
+        for mut next in drain {
+            if let Some(Frame::Kafka(KafkaFrame::Response {
+                body: ResponseBody::ListOffsets(next_list_offsets),
+                ..
+            })) = next.frame()
+            {
+                for next_topic in std::mem::take(&mut next_list_offsets.topics) {
+                    if let Some(base_response) = base_fetch
```
Comment (rukai) [inline shotover/src/transforms/kafka/sink_cluster/mod.rs:1732]: base_list_offsets

File: shotover/src/transforms/kafka/sink_cluster/mod.rs [hunk 3]
```diff
@@ -1576,6 +1718,46 @@ impl KafkaSinkCluster {
...
+                        .topics
+                        .iter_mut()
+                        .find(|topic| topic.name == next_topic.name)
+                    {
+                        for next_partition in &next_topic.partitions {
+                            for base_partition in &base_response.partitions {
```
Comment (rukai) [inline shotover/src/transforms/kafka/sink_cluster/mod.rs:1738]: base_topic

File: shotover/src/transforms/kafka/sink_cluster/mod.rs [hunk 4]
```diff
@@ -1548,6 +1698,7 @@ impl KafkaSinkCluster {
                 ))
             }
         }
+        base.invalidate_cache();
```
Comment (justinweng-instaclustr) [inline shotover/src/transforms/kafka/sink_cluster/mod.rs:1701]: why do we need 2 `base.invalidate_cache();` here?

**Review by ronycsdu — APPROVED** [review]

File: shotover/src/transforms/kafka/sink_cluster/mod.rs (hunk 4 above)
Comment (rukai) [reply shotover/src/transforms/kafka/sink_cluster/mod.rs]: good catch, its a bad merge, fixed

**Review by justinweng-instaclustr — APPROVED** [review]

**Pull request merged**